### PR TITLE
Add Expect: 100-continue pipeline policy to storage sdk

### LINKED ISSUE #20406: Add support for 100-continue in Azure Core
https://github.com/Azure/azure-sdk/issues/5787

### PR DESCRIPTION
**Pull request by syeleti-msft**
**Summary**
A new ExpectContinueBehavior field on ClientOptions (exposed in azblob, service, container, and blob) controls how the header is applied:

```
type ExpectContinueOptions struct {
    Mode                   ExpectContinueMode // ApplyOnThrottle (default), On, Off
    ContentLengthThreshold int64             // only attach when body size >= threshold
    ThrottleInterval           time.Duration     // throttle window for ApplyOnThrottle (default 1 min)
}
```
Modes:

ApplyOnThrottle (default) — attaches the header only for a window (AutoInterval, default 1 * time.Minute) after the pipeline sees a throttling response (429, 500, 503). Matches .NET's AutoInterval default.
On — always attach the header on requests with a known content length above the threshold.
Off — never attach.

Behavior also honors the AZURE_STORAGE_DISABLE_EXPECT_CONTINUE_HEADER environment variable (same name as .NET) — when truthy, no policy is added regardless of options.

**Timeout for 100-continue**
go SDK: 1s ([source](https://github.com/Azure/azure-sdk-for-go/blob/31da55da5e493c5b7a076dac90e182257ab82ac7/sdk/azcore/runtime/transport_default_http_client.go#L29))
dot net SDK: 1s ([source](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.expect100continuetimeout?view=net-10.0#:~:text=The-,timespan,-to%20wait%20for))

.NET SDK PR for reference: https://github.com/Azure/azure-sdk-for-net/pull/40611

- [x] The purpose of this PR is explained in this or a referenced issue.
- [ ] The PR does not update generated files.
   - These files are managed by the codegen framework at [Azure/autorest.go][].
- [x] Tests are included and/or updated for code changes.
- [ ] Updates to module CHANGELOG.md are included.
- [ ] MIT license headers are included in each file.

[Azure/autorest.go]: https://github.com/Azure/autorest.go

### CODE REVIEW

File: sdk/storage/azblob/internal/base/expect_continue.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+// header to outgoing requests according to the provided options.
+//
+// When opts is nil the default behavior is ExpectContinueModeApplyOnThrottle, equivalent to
+// passing &exported.ExpectContinueOptions{}.
+//
+// When the environment variable AZURE_STORAGE_DISABLE_EXPECT_CONTINUE_HEADER is set to a truthy
+// value, the returned policy is nil regardless of the supplied options.
+//
+// Returns nil when no policy should be added to the pipeline (mode is Off or env disables it).
+func NewExpectContinuePolicy(opts *exported.ExpectContinueOptions) policy.Policy {
```
[the remark this answers is not in the document]
Comment (syeleti-msft) [reply sdk/storage/azblob/internal/base/expect_continue.go]: modified the comment.

File: sdk/storage/azblob/internal/exported/expect_continue.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+	// ContentLengthThreshold is the minimum value of HTTP request Content-Length for applying
+	// the Expect: 100-continue header. The default is zero, meaning any request that has a
+	// non-empty body and a computable content length is eligible.
+	ContentLengthThreshold *int64
+
+	// AutoInterval is used in ExpectContinueModeApplyOnThrottle mode to set the time window
+	// during which the Expect: 100-continue header is applied after a triggering error response
+	// (429, 500, or 503) is observed. When nil, the default of one minute is used.
+	AutoInterval *time.Duration
```
Comment (tanyasethi-msft) [inline sdk/storage/azblob/internal/exported/expect_continue.go]: I think .NET uses ThrottleInterval — this would be more descriptive.

File: sdk/storage/azblob/internal/base/expect_continue_test.go [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/internal/exported"
+	"github.com/stretchr/testify/require"
+)
+
+// newPipelineForPolicy wraps the given policy in a pipeline targeting the supplied mock server.
+// Retries are disabled so each pipeline invocation produces exactly one transport call, making
+// expectation checks against the mock server's response sequence deterministic.
+func newPipelineForPolicy(p policy.Policy, srv policy.Transporter) runtime.Pipeline {
+	return runtime.NewPipeline("test", "v0.0.0",
+		runtime.PipelineOptions{PerCall: []policy.Policy{p}},
```
Comment (tanyasethi-msft) [inline sdk/storage/azblob/internal/base/expect_continue_test.go]: This builds the pipeline with PerCall: []policy.Policy{p}, but production adds it as PerRetry.
Doesn't cause bugs here (retries disabled) but it's an inconsistency.

File: sdk/storage/azblob/blockblob/client_test.go [hunk 4]
```diff
@@ -7037,3 +7047,62 @@ func (s *BlockBlobUnrecordedTestsSuite) TestCommitBlockListWithUDSCreatePermissi
...
+	require.NoError(t, err)
+
+	require.NotEmpty(t, observer.seen)
+	for _, v := range observer.seen {
+		require.Equal(t, "100-continue", v)
+	}
+}
+
+// TestUploadExpectContinueOff verifies the header is not set when the mode is Off.
+func TestUploadExpectContinueOff(t *testing.T) {
```
Comment (tanyasethi-msft) [inline sdk/storage/azblob/blockblob/client_test.go:7528]: lets test the default ApplyOnThrottle also behavior through an actual client.

File: sdk/storage/azblob/internal/base/expect_continue.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+	// expectContinueHeader is the name of the HTTP "Expect" header.
+	expectContinueHeader = "Expect"
+
+	// expectContinueHeaderValue is the value used when applying the "Expect: 100-continue" header.
+	expectContinueHeaderValue = "100-continue"
+
+	// DisableExpectContinueHeaderEnvVar is the name of the environment variable used to disable
+	// the automatic application of the "Expect: 100-continue" header. When the value parses to
+	// true (e.g. "1", "true", "TRUE"), no Expect: 100-continue policy will be added to the pipeline.
+	DisableExpectContinueHeaderEnvVar = "AZURE_STORAGE_DISABLE_EXPECT_CONTINUE_HEADER"
```
Comment (tanyasethi-msft) [inline sdk/storage/azblob/internal/base/expect_continue.go:28]: Lets add an integration test for this env var disable

**Review by tanyasethi-msft — COMMENTED** [review]
I believe .net and other had added this as a PerCall policy, but here we are adding it as PerRetry. This means the header gets re-evaluated on each retry attempt vs once before retries in .NET. Is this intentional?

**Comment by syeleti-msft** [comment]
> I believe .net and other had added this as a PerCall policy, but here we are adding it as PerRetry. This means the header gets re-evaluated on each retry attempt vs once before retries in .NET. Is this intentional?

Yes, but they changed this policy recently to per-retry. here is the PR: https://github.com/Azure/azure-sdk-for-net/pull/52438

File: sdk/storage/azblob/blockblob/client_test.go (hunk 4 above)
Comment (syeleti-msft) [reply sdk/storage/azblob/blockblob/client_test.go:7528]: added a mock test to test the default behaviour

File: sdk/storage/azblob/internal/base/expect_continue.go (hunk 5 above)
Comment (syeleti-msft) [reply sdk/storage/azblob/internal/base/expect_continue.go:28]: added to [sdk/storage/azblob/blockblob/client_test.go](https://github.com/Azure/azure-sdk-for-go/pull/26763/files#diff-0d6bcbed67db7eec034824c6cf97283e601cf8c592c982584d43541fdb67c5bc)

**Review by tanyasethi-msft — APPROVED** [review]

File: sdk/storage/azblob/internal/base/expect_continue.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+	// Accessed atomically to allow concurrent use across requests sharing a pipeline.
+	lastThrottleUnixNanos atomic.Int64
+}
+
+func (p *expectContinueOnThrottlePolicy) Do(req *policy.Request) (*http.Response, error) {
+	if shouldApplyExpectContinue(req, p.contentLengthThreshold) {
+		last := p.lastThrottleUnixNanos.Load()
+		if last != 0 && time.Now().UnixNano()-last < p.throttleInterval.Nanoseconds() {
+			req.Raw().Header.Set(expectContinueHeader, expectContinueHeaderValue)
+		}
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azblob/internal/base/expect_continue.go:130]: @syeleti-msft - can you look if this is a valid comment and this situation can occur in production if system time of the machine is changed?

**Comment by seanmcc-msft** [comment]
syeleti-msft, this feature also needs to be implemented for Files, Queues, and Data Lake.  See the .NET PR - https://github.com/Azure/azure-sdk-for-net/pull/40611/changes

File: sdk/storage/azblob/internal/base/clients.go [hunk 7]
```diff
@@ -20,6 +20,15 @@ type ClientOptions struct {
...
 	Audience string
+
+	// ExpectContinueBehavior configures the application of the HTTP "Expect: 100-continue"
+	// header on operations that include a request body. If nil, the default behavior is used,
+	// which conditionally applies the header for a short window after the service responds
+	// with a throttle/server-error status (429, 500, or 503).
+	//
+	// Setting the environment variable AZURE_STORAGE_DISABLE_EXPECT_CONTINUE_HEADER to a
+	// truthy value disables this behavior entirely, regardless of this setting.
+	ExpectContinueBehavior *exported.ExpectContinueOptions
```
Comment (jhendrixMSFT) [inline sdk/storage/azblob/internal/base/clients.go]: The preference is to have a meaningful zero-value instead of `nil` unless that results in ambiguity.  I think in this case it does not, so the `*` can be removed.

File: sdk/storage/azblob/internal/exported/expect_continue.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+type ExpectContinueOptions struct {
+	// Mode controls when the Expect: 100-continue header is applied. The default value
+	// (ExpectContinueModeApplyOnThrottle) applies the header for a fixed time window after
+	// the service responds with a throttle/server-error status (429, 500, or 503).
+	Mode ExpectContinueMode
+
+	// ContentLengthThreshold is the minimum value of HTTP request Content-Length for applying
+	// the Expect: 100-continue header. The default is zero, meaning any request that has a
+	// non-empty body and a computable content length is eligible.
+	ContentLengthThreshold *int64
```
Comment (jhendrixMSFT) [inline sdk/storage/azblob/internal/exported/expect_continue.go]: Can remove `*` here.

File: sdk/storage/azblob/internal/exported/expect_continue.go [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+	// ThrottleInterval is used in ExpectContinueModeApplyOnThrottle mode to set the time window
+	// during which the Expect: 100-continue header is applied after a triggering error response
+	// (429, 500, or 503) is observed. When nil, the default of one minute is used.
+	ThrottleInterval *time.Duration
```
Comment (jhendrixMSFT) [inline sdk/storage/azblob/internal/exported/expect_continue.go]: I think we should also get rid of the `*` here.  However, first I'd like to understand the allowed values here and what they mean.
- does zero have a special meaning here, i.e. no interval?
- are there stated lower and upper bounds?  e.g. allowing sub-second intervals seems dubious (see also https://github.com/Azure/azure-sdk-for-go/pull/26763/files#discussion_r3233557333)

File: sdk/storage/azblob/internal/base/expect_continue.go [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+	}
+	resp, err := req.Next()
+	if resp != nil {
```
Comment (jhendrixMSFT) [inline sdk/storage/azblob/internal/base/expect_continue.go:133]: This doesn't handle the case when the transport returns an error (i.e. `err != nil`).  Is that intentional?

File: sdk/storage/azblob/internal/base/expect_continue.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply sdk/storage/azblob/internal/base/expect_continue.go:130]: Claude gives the same feedback although it thinks the condition is more theoretical.  However, it does think that fixing it is cheap and worth doing.

File: sdk/storage/azblob/internal/base/expect_continue.go [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+	// lastThrottleUnixNanos stores the UnixNano timestamp of the most recent triggering response.
+	// Accessed atomically to allow concurrent use across requests sharing a pipeline.
+	lastThrottleUnixNanos atomic.Int64
+}
+
+func (p *expectContinueOnThrottlePolicy) Do(req *policy.Request) (*http.Response, error) {
+	if shouldApplyExpectContinue(req, p.contentLengthThreshold) {
+		last := p.lastThrottleUnixNanos.Load()
+		if last != 0 && time.Now().UnixNano()-last < p.throttleInterval.Nanoseconds() {
+			req.Raw().Header.Set(expectContinueHeader, expectContinueHeaderValue)
```
Comment (jhendrixMSFT) [inline sdk/storage/azblob/internal/base/expect_continue.go:127]: Note that if we add the header here it will persist even when the window expires.  It might not matter in practice, but the below would fix it.
```go
if shouldApplyExpectContinue(req, p.contentLengthThreshold) {
    last := p.lastThrottleUnixNanos.Load()
    if last != 0 && time.Now().UnixNano()-last < p.throttleInterval.Nanoseconds() {
        req.Raw().Header.Set(expectContinueHeader, expectContinueHeaderValue)
    } else {
        req.Raw().Header.Del(expectContinueHeader)
    }
}
```

File: sdk/storage/azblob/internal/exported/expect_continue.go (hunk 9 above)
Comment (syeleti-msft) [reply sdk/storage/azblob/internal/exported/expect_continue.go]: > does zero have a special meaning here, i.e. no interval?

the interval is only used in the mode `ExpectContinueModeApplyOnThrottle`. so zero value would would use the default one minute interval. 

added a comment in the code to describe the behaviour.
```
    // ThrottleInterval is used in ExpectContinueModeApplyOnThrottle mode to set the time window
    // during which the Expect: 100-continue header is applied after a triggering error response
    // (429, 500, or 503) is observed. A zero value uses the default of one minute.
```

no interval for a zero value doesn't make sense to me as it would behave like mode `ExpectContinueModeOff`

> are there stated lower and upper bounds? e.g. allowing sub-second intervals seems dubious (see also https://github.com/Azure/azure-sdk-for-go/pull/26763/files#discussion_r3233557333)

there are no such values as of now, actually the idea is to just expose this parameter in case, i think it would be hard for a user to come up with a value for this interval in general. I see .NET also don't have any such upper/lower bound for the interval.

File: sdk/storage/azblob/internal/base/expect_continue.go (hunk 11 above)
Comment (syeleti-msft) [reply sdk/storage/azblob/internal/base/expect_continue.go:127]: make sense, added it.

File: sdk/storage/azblob/internal/base/clients.go (hunk 7 above)
Comment (syeleti-msft) [reply sdk/storage/azblob/internal/base/clients.go]: okay, changed to a value field

File: sdk/storage/azblob/internal/exported/expect_continue.go (hunk 8 above)
Comment (syeleti-msft) [reply sdk/storage/azblob/internal/exported/expect_continue.go]: changed to value field.

File: sdk/storage/azblob/internal/base/expect_continue.go (hunk 10 above)
Comment (syeleti-msft) [reply sdk/storage/azblob/internal/base/expect_continue.go:133]: as per my understanding, the response status code is enough to take the decision on enabling the `expectContinueOnThrottlePolicy`. I don't see how err can help us here

File: sdk/storage/azblob/internal/base/expect_continue.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (syeleti-msft) [reply sdk/storage/azblob/internal/base/expect_continue.go:130]: agreed, fixed by converting the throttleInterval to  atomic.Pointer[time.Time]

File: sdk/storage/azblob/internal/base/expect_continue.go (hunk 10 above)
Comment (jhendrixMSFT) [reply sdk/storage/azblob/internal/base/expect_continue.go:133]: The underlying HTTP transport will return an error e.g. if the network connection is dropped/timed out.  Could this also happen under throttling?  It might be ambiguous though as it could drop for non-throttling reasons too.  Do the other languages only apply this header based on the HTTP status code?  If so, then there's no changes required.

File: sdk/storage/azblob/internal/exported/expect_continue.go (hunk 9 above)
Comment (jhendrixMSFT) [reply sdk/storage/azblob/internal/exported/expect_continue.go]: I see that the `*` was removed, so a zero-value will mean use the default (this is good).

> i think it would be hard for a user to come up with a value for this interval in general

Do we think this is worth making public then?  Mostly my concern is around bugs showing up due to edge cases that haven't been well tested.  e.g. we have seen cases where customers incorrectly pass `1` for a `time.Duration` thinking it means one second.

At any rate, the concern isn't blocking, just making sure we're thinking about these edge cases.  If there's nothing obviously broken with small/large bounds then we can leave it as is.

**Review by jhendrixMSFT — APPROVED** [review]

File: sdk/storage/azblob/internal/exported/expect_continue.go (hunk 9 above)
Comment (syeleti-msft) [reply sdk/storage/azblob/internal/exported/expect_continue.go]: >Do we think this is worth making public then?

Yes, we should expose this to align with .NET SDK approach.
I see no perf issue to hit when user specifies small interval as it would prevent 100continue header to be applied on the request in throttle.

File: sdk/storage/azblob/internal/base/expect_continue.go (hunk 10 above)
Comment (syeleti-msft) [reply sdk/storage/azblob/internal/base/expect_continue.go:133]: >The underlying HTTP transport will return an error e.g. if the network connection is dropped/timed out. Could this also happen under throttling?

it is written in spec doc that "The service may choose to close connections while attempting to defend itself against an aggressive client. Especially if consuming request bodies and sending back throttling response isn’t sustainable. Therefore, elevated connection reset rate should prompt 100-Continue." 

> Do the other languages only apply this header based on the HTTP status code?

I see .NET has implemented it with status codes https://github.com/Azure/azure-sdk-for-net/blob/7ff7e4bc2b94eb2f82c7c5ac6a1eb50eaf38bc3c/sdk/storage/Azure.Storage.Common/src/Shared/ExpectContinueOnThrottlePolicy.cs#L39

File: sdk/storage/azblob/internal/exported/expect_continue.go (hunk 9 above)
Comment (jhendrixMSFT) [reply sdk/storage/azblob/internal/exported/expect_continue.go]: It's not a question of perf, it's if small/large values will cause incorrect behavior.  Are there unit tests that cover small/large values?

File: sdk/storage/azblob/internal/base/expect_continue.go (hunk 10 above)
Comment (jhendrixMSFT) [reply sdk/storage/azblob/internal/base/expect_continue.go:133]: While I can't speak for .NET, in Go, if the peer resets the connection, then the transport will return an error (there's no `*http.Response` as the peer didn't send one).  So, if we need to enable `100-Continue` on connection errors then we'd want to include non `nil` errors to the condition.

**Comment by jhendrixMSFT** [comment]
The CI failure is a known issue with the test proxy; I'm working on getting it resolved.

**Comment by seanmcc-msft** [comment]
> syeleti-msft, this feature also needs to be implemented for Files, Queues, and Data Lake. See the .NET PR - https://github.com/Azure/azure-sdk-for-net/pull/40611/changes

Resolved offline, .NET only supports 100-continue for Blobs.

File: sdk/storage/azblob/internal/base/expect_continue.go (hunk 10 above)
Comment (syeleti-msft) [reply sdk/storage/azblob/internal/base/expect_continue.go:133]: this PR is intended to replicate the behavior of what .NET does today, while I agree with you that non `nil` errors should be considered in the condition, but the heuristic to decide how many number of connections resets are needed to trigger 100-continue is something yet to be decided and it can be done in a new PR. 

"there are valid scenarios where connection reset can occur (for example during larger read where data is being changed concurrently and FE can’t maintain consistent read but already sent response headers)." I think we can't naively apply 100-continue header for every connection reset error as there were valid scenarios.

File: sdk/storage/azblob/internal/exported/expect_continue.go (hunk 9 above)
Comment (syeleti-msft) [reply sdk/storage/azblob/internal/exported/expect_continue.go]: added the unit test for these edge cases in this commit https://github.com/Azure/azure-sdk-for-go/pull/26763/changes/e76002fe82e1b993b38b66c6a6fdfbae13e61436

**Review by jhendrixMSFT — APPROVED** [review]

**Review by jhendrixMSFT — APPROVED** [review]

**Review by souravgupta-msft — APPROVED** [review]
